packages/xo-server/src/api/vm.mjs

```diff
@@ -48,6 +48,44 @@ const extract = (obj, prop) => {
   return value
 }
 
+const startVmAndDestroyCloudConfigVdi = async (xapi, vm, vdiUuid, params) => {
+  try {
+    const start = new Date()
+    await xapi.startVm(vm._xapiId)
+
+    if (params.destroyCloudConfigVdiAfterBoot && vdiUuid !== undefined) {
+      const started = new Date()
+      const timeout = 10 * 60 * 1000
+      const startDuration = started - start
+
+      let remainingTimeout = timeout - startDuration
+
+      // wait for the 'Running' event to be really stored in local xapi object cache
+      await xapi.waitObjectState(vm.$ref, vm => vm.power_state === 'Running', {
```

Comment:
`vm` is a XO object, not a XAPI object, and `$ref` is only defined on XAPI objects.

The problem may be similar for other things in this function, choose either one or the other but avoid mixing them.